imitation/tests/test_tabular.py

```diff
@@ -0,0 +1,261 @@
+"""Test tabular environments and tabular MCE IRL."""
+
+import numpy as np
+
+from imitation.model_env import ModelBasedEnv, RandomMDP
+from imitation.tabular_irl import (SGD, AMSGrad, maxent_irl,
+                                   mce_occupancy_measures, mce_partition_fh,
+                                   LinearRewardModel)
+
+# import pytest
+
+
+def rollouts(env, n=10, seed=None):
+    rv = []
+    for i in range(n):
+        done = False
+        if seed is not None:
+            # if a seed is given, then we use the same seed each time (should
+            # give same trajectory each time)
+            env.seed(seed)
+            env.action_space.seed(seed)
+        obs = env.reset()
+        traj = [obs]
+        acts = []
+        while not done:
+            act = env.action_space.sample()
+            acts.append(act)
+            obs, rew, done, info = env.step(act)
+            traj.append((obs, rew))
+        rv.append(traj)
+    return rv
+
+
+def test_random_mdp():
+    for i in range(3):
+        n_states = 4 * (i + 3)
+        n_actions = i + 2
+        branch_factor = i + 1
+        if branch_factor == 1:
+            # make sure we have enough actions to get reasonable trajectories
+            n_actions = min(n_states, max(n_actions, 4))
+        horizon = 5 * (i + 1)
+        random_obs = (i % 2) == 0
+        obs_dim = (i * 3 + 4)**2 + i
+        mdp = RandomMDP(
+            n_states=n_states,
+            n_actions=n_actions,
+            branch_factor=branch_factor,
+            horizon=horizon,
+            random_obs=random_obs,
+            obs_dim=obs_dim if random_obs else None,
+            generator_seed=i)
+
+        # sanity checks on sizes of things
+        assert mdp.transition_matrix.shape == (n_states, n_actions, n_states)
+        assert mdp.observation_matrix.shape[0] == n_states \
+            and mdp.observation_matrix.ndim == 2
+
+        # make sure trajectories aren't all the same if we don't specify same
+        # seed each time
+        trajectories = rollouts(mdp, 100)
+        assert len(set(map(str, trajectories))) > 1
+
+        trajectories = rollouts(mdp, 100, seed=42)
+        # make sure trajectories ARE all the same if we do specify the same
+        # seed each time
+        assert len(set(map(str, trajectories))) == 1
+
+
+def test_policy_om_random_mdp():
+    mdp = RandomMDP(
+        n_states=16,
+        n_actions=3,
+        branch_factor=2,
+        horizon=20,
+        random_obs=True,
+        obs_dim=5,
+        generator_seed=42)
+    V, Q, pi = mce_partition_fh(mdp)
+    assert np.all(np.isfinite(V))
+    assert np.all(np.isfinite(Q))
+    assert np.all(np.isfinite(pi))
+    assert np.all(pi >= 0)
+    # it (always?) has to take SOME actions
+    assert np.any(pi > 0)
+    # we do <= 1 instead of allclose() because pi actually doesn't have to be
```

Comment:
Oop, you're right. This comment was for a previous version of the code that iteratively applied Ziebart's infinite horizon soft backup (Algorithm 9.1 in his thesis; see bottom of p.111 for exposition), which sometimes doesn't converge to a properly normalised distribution because it incorporates a "termination reward" that allows (and potentially encourages) the agent to terminate early. The current implementation here definitely doesn't do that.